### PR TITLE
[MERGE] crm(_*): move from onchange to editable stored fields

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
PURPOSE

Try to move from onchange / default_get to stored editable computed fields.
Behavior should be the same (computed or set by user), with support of
create / write / onchange field update without additional code.

USE EDITABLE STORED FIELDS

Update classic fields updated in some cases by onchange and/or default methods
by fields with store=True, readonly=False. It means their value comes either
from manual user input, either from trigger based computation.

Remove onchange and default_get when possible, leading to an unique computation
method and clearing fields definition.

Also clean some fields definition inconsistencies, notably required fields
that should instead be correctly computed or default that have no real meaning.

CLEAN CRM CONVERT WIZARDS CODE AND FIELDS USE

In this merge we also rewrite code located and/or used in convert or merge
wizards. Indeed it holds unnecessary variables, dictionaries manipulation
that hides the real business flow, badly named variables, unclear methods, ...
Logic of those wizards seem quite simple but is hidden behind a wall of
bloated code.

CLEAN CRM.TEAM MODEL

Clean statistics fields definition on crm.team. Future tasks will improve
the use of sales team in crm and first steps is to clean a bit models linked
to sales team.

LINKS
Task ID 2088565 (crm: from onchange to compute)
Community PR odoo/odoo#41955
Enterprise PR odoo/enterprise#7322
Upgrade PR odoo/upgrade#781

### CODE REVIEW

File: addons/crm/wizard/crm_lead_to_opportunity_mass_views.xml [hunk 1]
```diff
@@ -56,7 +56,6 @@
         res_model="crm.lead2opportunity.partner.mass"
         binding_model="crm.lead" binding_views="list"
         view_mode="form" target="new"
-        context="{'mass_convert' : True}"
```
Comment (KangOl) [inline addons/crm/wizard/crm_lead_to_opportunity_mass_views.xml]: Please force empty value to help upgrade

File: addons/crm/models/crm_lead.py [hunk 2]
```diff
@@ -230,6 +264,34 @@ def _compute_email_state(self):
...
+    def _compute_probabilities(self):
+        for lead in self:
+            was_automated = False
+            lead_probabilities = lead._pls_get_naive_bayes_probabilities()
+            if lead.id in lead_probabilities:
+                if tools.float_compare(lead.probability, lead.automated_probability, 2) == 0 and lead.active:
+                    was_automated = True
+                lead.automated_probability = lead_probabilities[lead.id]
+                if was_automated:
+                    lead.probability = lead.automated_probability
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py]: why not re-using is_automated_probability ?

File: addons/crm/models/crm_lead.py [hunk 3]
```diff
@@ -452,7 +402,7 @@ def _update_probability(self):
             if lead.id in lead_probabilities:
                 lead_proba = lead_probabilities[lead.id]
                 proba_vals = {'automated_probability': lead_proba}
-                if tools.float_compare(lead.probability, lead.automated_probability, 2) == 0:
+                if tools.float_compare(lead.probability, lead.automated_probability, 2) == 0 and lead.active:
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py]: same here: why not using is_automated_probability ?

File: addons/crm/models/crm_lead.py [hunk 4]
```diff
@@ -537,7 +487,7 @@ def _stage_find(self, team_id=False, domain=None, order='sequence'):
 
     def toggle_active(self):
         res = super(Lead, self).toggle_active()
-        self.filtered(lambda lead: lead.active)._update_probability()
+        self.filtered(lambda lead: lead.active)._compute_probabilities()
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py]: why this change ?

File: addons/crm/models/crm_lead.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/crm/models/crm_lead.py]: Indeed.

File: addons/crm/models/crm_lead.py (hunk 4 above)
Comment (tde-banana-odoo) [reply addons/crm/models/crm_lead.py]: Don't know. Reverted :) .

File: addons/crm/models/crm_lead.py (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/crm/models/crm_lead.py]: Why ? Why not ? Am I a dog ?

I am not a dog. I will use is_automated_probability .

Or am I a dog ?

Not sure...

Coin ...

Not a dog. Definitively.

File: addons/crm/models/crm_lead.py [hunk 5]
```diff
@@ -870,34 +870,6 @@ def opps_key(opportunity):
 
         return self.sorted(key=opps_key, reverse=reverse)
 
-    def get_duplicated_leads(self, partner_id, include_lost=False):
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py]: Maybe you can add in commit message that this method was only used in enterprise module (to explain this is why we cannot find any usage of it in this commit)

File: addons/crm/models/crm_lead.py [hunk 6]
```diff
@@ -967,106 +989,60 @@ def _create_lead_partner_data(self, partner_name, is_company=False, parent_id=Fa
...
-            partner_ids[lead.id] = partner_id
-        return partner_ids
-
-    def allocate_salesman(self, user_ids=None, team_id=False):
+            if force_partner_id:
+                lead.partner_id = force_partner_id
+            if not lead.partner_id and create_missing:
+                partner = lead._create_customer()
+                lead.partner_id = partner.id
+        return True
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py]: its return True if everything went ok I guess. So it should return false if something went wrong ?
if not, why return something ? Especially, here, the returned value is never used (even in _convert_handle_partner, the return value is useless)

File: addons/crm/models/crm_lead.py [hunk 7]
```diff
@@ -967,106 +989,60 @@ def _create_lead_partner_data(self, partner_name, is_company=False, parent_id=Fa
...
             if user_ids:
-                value['user_id'] = user_ids[index]
-                # Cycle through user_ids
-                index = (index + 1) % len(user_ids)
-            if value:
-                lead.write(value)
+                update_vals['user_id'] = user_ids[index % len(user_ids)]
+            if update_vals:
+                lead.write(update_vals)
         return True
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py]: Same here, I don't think it's useful to return something as the return value is never used

File: addons/crm/tests/test_crm_lead_convert.py [hunk 8]
```diff
@@ -8,6 +8,10 @@
 
 @tagged('lead_manage')
 class TestLeadConvert(crm_common.TestLeadConvertCommon):
+    """
+    TODO: created partner (handle assignation) has team of lead
+    TODO: create partner has user_id  coming from wizard
```
Comment (dbeguin) [inline addons/crm/tests/test_crm_lead_convert.py]: allez, do it so it's behind us after :D

File: addons/crm/wizard/crm_lead_to_opportunity.py [hunk 9]
```diff
@@ -81,79 +81,63 @@ def _onchange_user(self):
...
+        if not result_opportunity.active:
+            result_opportunity.write({'active': True, 'activity_type_id': False, 'lost_reason': False})
+
+        if result_opportunity.type == "lead":
+            self._convert_and_allocate(result_opportunity, [self.user_id.id], team_id=self.team_id.id)
+        elif not self._context.get('no_force_assignation') or not result_opportunity.user_id:
+            result_opportunity.write({
+                'user_id': self.user_id.id,
+                'team_id': self.team_id.id,
+            })
```
Comment (dbeguin) [inline addons/crm/wizard/crm_lead_to_opportunity.py]: while we are at it, could we instead prepare a value dict and write everything at the end if needed ?

File: addons/crm/models/crm_lead.py [hunk 10]
```diff
@@ -916,17 +916,19 @@ def _get_lead_duplicates(self, partner=None, email=None, include_lost=False):
...
         """
-        if not email:
+        if not email and not partner:
             return self.env['crm.lead']
-        partner_match_domain = []
-        for email in set(email_split(email) + [email]):
-            partner_match_domain.append(('email_from', '=ilike', email))
+
+        domain = []
+        for normalized_email in [tools.email_normalize(email) for email in tools.email_split(email)]:
```
Comment (dbeguin) [inline addons/crm/models/crm_lead.py]: Just a quite old thinking :
I'm wondering at what point we should make email_normalize = a string list of email normalized.
We could do for email in email_split(self.email_normalized) directly, everywhere.
But it could become super annoying to use.. for Now, I think we consider the email_normalize field contain only the first email of the email string list. (because email field can contain multiple email addresses).
There is something to do somehow here I think.. But maybe not today..

Comment (dbeguin) [reply addons/crm/models/crm_lead.py]: why not keeping `for normalized_email in **set**(tools.email_normalize(email) for email in tools.email_split(email))` ?

File: addons/crm/wizard/crm_lead_to_opportunity.py [hunk 11]
```diff
@@ -55,10 +55,14 @@ def default_get(self, fields):
         ('nothing', 'Do not link to a customer')
     ], string='Related Customer', required=True)
     lead_id = fields.Many2one('crm.lead', "Associated Lead")
-    opportunity_ids = fields.Many2many('crm.lead', string='Opportunities')
+    duplicated_lead_ids = fields.Many2many(
+        'crm.lead', string='Duplicates', context={'active_test': False})
```
Comment (dbeguin) [inline addons/crm/wizard/crm_lead_to_opportunity.py]: Nice one!

File: addons/crm/wizard/crm_lead_to_opportunity_mass.py [hunk 12]
```diff
@@ -29,7 +28,7 @@ def default_get(self, fields):
     action = fields.Selection(selection_add=[
         ('each_exist_or_create', 'Use existing partner or create'),
     ], string='Related Customer', required=True)
-    force_assignation = fields.Boolean('Force assignation', help='If unchecked, this will leave the salesman of duplicated opportunities')
+    force_assignemnt = fields.Boolean(default=False)
```
Comment (dbeguin) [inline addons/crm/wizard/crm_lead_to_opportunity_mass.py]: why remove the help ?
Also, don't we prefer to always add a string for every field ?

File: addons/crm/wizard/crm_lead_to_opportunity_views.xml [hunk 13]
```diff
@@ -5,6 +5,7 @@
         <field name="model">crm.lead2opportunity.partner</field>
         <field name="arch" type="xml">
             <form string="Convert to Opportunity">
+                <field name="force_assignment" invisible="1"/>
```
Comment (dbeguin) [inline addons/crm/wizard/crm_lead_to_opportunity_views.xml]: is it used somewhere ? or it's added to the view to have it in memory during the wizard execution ? :thinking:

File: addons/website_crm_partner_assign/models/crm_lead.py [hunk 14]
```diff
@@ -18,22 +19,25 @@ class CrmLead(models.Model):
...
-            self.date_assign = False
-        else:
-            self.date_assign = fields.Date.context_today(self)
-            self.user_id = partner_assigned.user_id
+        for lead in self:
+            if not lead.partner_assigned_id:
+                lead.date_partner_assign = False
+            else:
+                lead.date_assign = fields.Date.context_today(lead)
+            # self.user_id = partner_assigned.user_id
```
Comment (dbeguin) [inline addons/website_crm_partner_assign/models/crm_lead.py]: remove this

File: addons/crm/models/crm_lead.py (hunk 6 above)
Comment (tde-banana-odoo) [reply addons/crm/models/crm_lead.py]: Done

File: addons/crm/models/crm_lead.py (hunk 7 above)
Comment (tde-banana-odoo) [reply addons/crm/models/crm_lead.py]: Ok done

File: addons/crm/tests/test_crm_lead_convert.py (hunk 8 above)
Comment (tde-banana-odoo) [reply addons/crm/tests/test_crm_lead_convert.py]: Plein le luc comme on dit. Ca sera pour plus tarte.

File: addons/crm/wizard/crm_lead_to_opportunity_views.xml (hunk 13 above)
Comment (tde-banana-odoo) [reply addons/crm/wizard/crm_lead_to_opportunity_views.xml]: Indeed not used ... could be removed. It is not in any triggers.

File: addons/crm/wizard/crm_lead_to_opportunity_mass.py (hunk 12 above)
Comment (tde-banana-odoo) [reply addons/crm/wizard/crm_lead_to_opportunity_mass.py]: Actually it comes from inherit. It has been added on crm_lead2opportunity model that is inherited from this one.

File: addons/crm/models/crm_lead.py (hunk 10 above)
Comment (tde-banana-odoo) [reply addons/crm/models/crm_lead.py]: I adivse you to note that in a task so that we can have a chat about it afterwards ... this one begins to give me boutons.

File: addons/crm/wizard/crm_lead_to_opportunity.py (hunk 9 above)
Comment (tde-banana-odoo) [reply addons/crm/wizard/crm_lead_to_opportunity.py]: Actually I have another task which will remove the first write and call action_unarchive instead :D . So there won't be two consecutive write anymore.

File: addons/crm/wizard/crm_lead_to_opportunity.py (hunk 11 above)
Comment (tde-banana-odoo) [reply addons/crm/wizard/crm_lead_to_opportunity.py]: :heart:

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
